doc/manual/contract.md

```diff
@@ -0,0 +1,535 @@
+
+All contracts conditions can be declared in several ways, all are equivalent:
+~~~nitish
+fun foo(a: Int, b: Int) is expect(a > 0, b < 0)
+~~~
+~~~nitish
+fun foo(a: Int, b: Int) is expect(a > 0 and b < 0)
+~~~
+~~~nitish
+fun foo(a: Int, b: Int) is expect(a > 0), ensure(b < 0)
+~~~
+
+## Writing invariants
+
+Invariants are used to specify the characteristics of a class/interface which must always be true, except when executing a member function. In other words, invariants define conditions that must be respected by our instance. Any instance which does not respect one or more of those conditions will be considered as in an incoherent state, and will cause an "assertion failed" error at runtime.
+
+To define an invariant we use the class annotation `invariant`. Here is an example with our Stack. The class defines that it is impossible to be in a state where the size is less than 0.
+
+~~~nitish
+class Stack
+	invariant(size >= 0)
+end
+~~~
+
+Another example with a class that represents a date (day, month). To be in a consistent state, the date must be represented with a day between 1 and 31 and a month between 1 and 12.
+
+~~~
+class Date
+	invariant(day >= 1 and day <= 31)
+	invariant(month >= 1 and month <= 12)
+
+	var day: Int
+	var month: Int
+end
+~~~
+
+When a class defines an invariant all methods and constructors (inherited, redefined and introduced) will check this one. By default, the invariants are only checked on exit. See section `Verification policy` to activate them on in and out. For all constructors the invariant is only checked at the end of this, this is due to the fact that the object is not yet initialized. See the section `Execution order` for example.
+
+The invariant is not checked when the call is on `self`. The following example shows the execution of invariant:
+
+~~~
+class A
+	invariant(print_invariant)
+
+	fun print_invariant: Bool
+	do
+		print "In invariant verification"
+		return true
+	end
+
+	fun foo
+	do
+		print "In foo"
+		self.bar # The called property is a member of "self", the invariant will not be checked.
+	end
+
+	fun bar
+	do
+		print "In bar"
+	end
+end
+
+var a = new A
+a.foo # The called property is not a member of "self", the invariant will be checked.
+~~~
+
+This example gives the following output:
+~~~nitish
+In invariant verification
+In foo
+In bar
+In invariant verification
+~~~
+
+## Writing contracts
+
+Methods, as well as attribute setters, can be associated with two types of contracts; precondition (`expect`) that expresses requirements of the routine´s call and postcondition (`ensure`) that expresses what the routine guarantee after her execution. Together, they define the rules to be respected between the callee and the caller. If the caller fulfills the `expect` condition on entry then the caller can assume that the callee fulfill the `ensure` condition on exit.
```

Comment:
Il y a deux erreurs de grammaire:
If the caller fulfills the `expect` condition on entry then the caller can assume that the callee fulfill the `ensure` condition on exit.
-> 
If the caller fulfills the `expect` condition on entry**,** then the caller can assume that the callee **fulfills** the `ensure` condition on exit.